Flaky test: TestExecute_HelloPythonSucceeds depends on live httpbin.org

## Problem

`TestExecute_HelloPythonSucceeds` (`pkg/runtime/python/runtime_405_test.go`) fails intermittently in CI when the external service it calls is unavailable. Observed on PR #413's CI run:

```
--- FAIL: TestExecute_HelloPythonSucceeds (11.32s)
    runtime_405_test.go:136: hello-python failed: exit status 1
    [warn] httpx.HTTPStatusError: Server error '503 Service Temporarily
           Unavailable' for url 'https://httpbin.org/get?name=World'
FAIL  github.com/dicode/dicode/pkg/runtime/python
```

The test runs the on-disk `tasks/examples/hello-python` task, which makes an allowlisted `httpx` GET to `https://httpbin.org`. When httpbin returns 5xx or is unreachable, the task exits non-zero and the test hard-fails — even though nothing in dicode is broken. This is unrelated to whatever change triggered the CI run (it surfaced on a Deno-only PR).

## Why the external call exists

The test deliberately exercises three things together: async `def main()` invocation (the #405 kwargs bug), PEP 723 / `uv` dependency provisioning of `httpx`, and the PEP 578 net guard admitting an **allowlisted** outbound host. The last point is why it reaches the public internet rather than mocking.

## Fix options

1. **Local `httptest.Server` (preferred).** Stand up an in-process HTTP server, allowlist its host (`127.0.0.1`), and point the task at it. Removes the external dependency entirely while still validating the net guard against a real allowlisted host. Requires making the example task's target URL configurable (param/env) so the test can override it — the task currently hardcodes the httpbin URL.
2. **Tolerate transient upstream failures.** If the task error is a 5xx / connection failure to the external host, `t.Skip` instead of `t.Fatal`. Cheaper, but weakens coverage when httpbin is down and leaves CI dependent on a third party.

Recommend option 1: it makes the test hermetic and keeps full coverage. Option 2 is an acceptable stopgap.

## Scope note

Surfaced while reviewing PR #413 (`env_read_exposed`); intentionally **not** fixed there to keep that PR scoped to the Deno env-permission change.

--- a/pkg/runtime/python/runtime_405_test.go
+++ b/pkg/runtime/python/runtime_405_test.go
@@ -4,7 +4,9 @@ package python
 
 import (
 	"context"
+	"encoding/json"
 	"net/http"
+	"net/http/httptest"
 	"strings"
 	"testing"
 	"time"
@@ -87,29 +89,34 @@ func TestExecute_FailedRunCapturesDiagnostics(t *testing.T) {
 
 // TestExecute_HelloPythonSucceeds is the end-to-end regression for #405: the
 // hello-python example (async def main(), PEP 723 httpx dep, an allowlisted
-// httpx GET to httpbin.org) must run to completion. It provisions uv the way
-// the runtime itself does — uv is not assumed to be on PATH — and reaches the
-// public internet, so it skips when uv cannot be provisioned. It guards the
-// async-main invocation (the kwargs bug that broke the run) and confirms the
-// PEP 578 net guard admits an allowlisted httpx call.
+// httpx GET) must run to completion. It provisions uv the way the runtime
+// itself does and skips when uv cannot be provisioned.
+//
+// The test spins up a local httptest.Server instead of reaching httpbin.org,
+// eliminating the external dependency that caused CI flakiness (#422). The
+// local server still validates the PEP 578 net guard: IP literals (127.0.0.1)
+// are allowed by the guard regardless of the allowlist, so the test uses
+// permissions.net: ["127.0.0.1"] to engage allowlist mode rather than
+// unrestricted mode while still permitting the loopback connection.
 func TestExecute_HelloPythonSucceeds(t *testing.T) {
 	if testing.Short() {
-		t.Skip("skipping network/uv integration test in -short mode")
+		t.Skip("skipping uv integration test in -short mode")
 	}
 	uv, err := uvpkg.EnsureUv("")
 	if err != nil {
 		t.Skipf("uv provisioning failed (offline?): %v", err)
 	}
 
-	// Pre-flight: skip if httpbin.org is not returning success (rate-limit, block, etc.).
-	pf, pfErr := http.Get("https://httpbin.org/get") //nolint:noctx
-	if pfErr != nil {
-		t.Skipf("httpbin.org unreachable: %v", pfErr)
-	}
-	pf.Body.Close()
-	if pf.StatusCode >= 300 {
-		t.Skipf("httpbin.org returned %d, skipping network integration test", pf.StatusCode)
-	}
+	// Local httpbin-compatible server — no external network required.
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+		_ = json.NewEncoder(w).Encode(map[string]any{
+			"origin": "127.0.0.1",
+			"args":   r.URL.Query(),
+			"url":    r.URL.String(),
+		})
+	}))
+	t.Cleanup(srv.Close)
 
 	rt, reg := newTestRuntime(t)
 	ex := rt.NewExecutor(uv)
@@ -124,8 +131,12 @@ func TestExecute_HelloPythonSucceeds(t *testing.T) {
 		Params: []task.Param{
 			{Name: "name", Default: "World"},
 			{Name: "count", Default: "1"},
+			{Name: "httpbin_url", Default: "https://httpbin.org/get"},
 		},
-		Permissions: task.Permissions{Net: []string{"httpbin.org"}},
+		// 127.0.0.1 is an IP literal — the PEP 578 guard admits it in
+		// allowlist mode without needing an explicit allowlist entry. Using
+		// a non-empty net list here engages allowlist mode rather than deny.
+		Permissions: task.Permissions{Net: []string{"127.0.0.1"}},
 	}
 	if err := reg.Register(spec); err != nil {
 		t.Fatalf("register: %v", err)
@@ -137,7 +148,10 @@ func TestExecute_HelloPythonSucceeds(t *testing.T) {
 		t.Fatalf("StartRun: %v", err)
 	}
 
-	res, err := ex.Execute(ctx, spec, pkgruntime.RunOptions{RunID: runID})
+	res, err := ex.Execute(ctx, spec, pkgruntime.RunOptions{
+		RunID:  runID,
+		Params: map[string]string{"httpbin_url": srv.URL + "/get"},
+	})
 	if err != nil {
 		t.Fatalf("Execute: %v", err)
 	}

--- a/pkg/runtime/python/runtime_405_test.go
+++ b/pkg/runtime/python/runtime_405_test.go
@@ -92,12 +92,17 @@ func TestExecute_FailedRunCapturesDiagnostics(t *testing.T) {
 // httpx GET) must run to completion. It provisions uv the way the runtime
 // itself does and skips when uv cannot be provisioned.
 //
-// The test spins up a local httptest.Server instead of reaching httpbin.org,
-// eliminating the external dependency that caused CI flakiness (#422). The
-// local server still validates the PEP 578 net guard: IP literals (127.0.0.1)
-// are allowed by the guard regardless of the allowlist, so the test uses
-// permissions.net: ["127.0.0.1"] to engage allowlist mode rather than
-// unrestricted mode while still permitting the loopback connection.
+// The test spins up a local httptest.Server and injects its URL via the
+// HTTPBIN_URL env var (forwarded to the subprocess via permissions.env),
+// eliminating the external dependency that caused CI flakiness (#422). Using
+// an env var rather than a run param means the URL is controlled only by the
+// host environment — it cannot be overridden by a task caller via the API.
+//
+// Net guard: the spec uses permissions.net: ["httpbin.org"] (matching the
+// production task.yaml). In allowlist mode the PEP 578 guard unconditionally
+// admits IP literals at the _is_ip_literal check before any hostname
+// comparison, so the 127.0.0.1 loopback address is allowed even though it
+// doesn't appear in the declared list.
 func TestExecute_HelloPythonSucceeds(t *testing.T) {
 	if testing.Short() {
 		t.Skip("skipping uv integration test in -short mode")
@@ -110,13 +115,20 @@ func TestExecute_HelloPythonSucceeds(t *testing.T) {
 	// Local httpbin-compatible server — no external network required.
 	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		w.Header().Set("Content-Type", "application/json")
-		_ = json.NewEncoder(w).Encode(map[string]any{
+		if err := json.NewEncoder(w).Encode(map[string]any{
 			"origin": "127.0.0.1",
 			"args":   r.URL.Query(),
 			"url":    r.URL.String(),
-		})
+		}); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
 	}))
 	t.Cleanup(srv.Close)
+
+	// Point the task at the local server via the host env var. The runtime
+	// forwards HTTPBIN_URL to the Python subprocess because it is declared
+	// under permissions.env in the spec below.
+	t.Setenv("HTTPBIN_URL", srv.URL+"/get")
 
 	rt, reg := newTestRuntime(t)
 	ex := rt.NewExecutor(uv)
@@ -131,12 +143,11 @@ func TestExecute_HelloPythonSucceeds(t *testing.T) {
 		Params: []task.Param{
 			{Name: "name", Default: "World"},
 			{Name: "count", Default: "1"},
-			{Name: "httpbin_url", Default: "https://httpbin.org/get"},
 		},
-		// 127.0.0.1 is an IP literal — the PEP 578 guard admits it in
-		// allowlist mode without needing an explicit allowlist entry. Using
-		// a non-empty net list here engages allowlist mode rather than deny.
-		Permissions: task.Permissions{Net: []string{"127.0.0.1"}},
+		Permissions: task.Permissions{
+			Net: []string{"httpbin.org"},
+			Env: []task.EnvEntry{{Name: "HTTPBIN_URL"}},
+		},
 	}
 	if err := reg.Register(spec); err != nil {
 		t.Fatalf("register: %v", err)
@@ -148,10 +159,7 @@ func TestExecute_HelloPythonSucceeds(t *testing.T) {
 		t.Fatalf("StartRun: %v", err)
 	}
 
-	res, err := ex.Execute(ctx, spec, pkgruntime.RunOptions{
-		RunID:  runID,
-		Params: map[string]string{"httpbin_url": srv.URL + "/get"},
-	})
+	res, err := ex.Execute(ctx, spec, pkgruntime.RunOptions{RunID: runID})
 	if err != nil {
 		t.Fatalf("Execute: %v", err)
 	}

--- a/tasks/examples/hello-python/task.py
+++ b/tasks/examples/hello-python/task.py
@@ -8,6 +8,7 @@
 #   - httpx async HTTP call as a realistic async dependency example
 
 import httpx
+import os
 
 
 async def main():
@@ -24,9 +25,7 @@ async def main():
     await kv.set_async("previous_name", name)
 
     # Async HTTP call with httpx (PEP 723 inline dep above).
-    # httpbin_url defaults to https://httpbin.org/get but can be overridden
-    # via params (e.g. by tests using a local httptest server).
-    httpbin_url = await params.get_async("httpbin_url", "https://httpbin.org/get")
+    httpbin_url = os.environ.get("HTTPBIN_URL", "https://httpbin.org/get")
     async with httpx.AsyncClient(timeout=5) as client:
         resp = await client.get(httpbin_url, params={"name": name})
         resp.raise_for_status()

--- a/tasks/examples/hello-python/task.py
+++ b/tasks/examples/hello-python/task.py
@@ -24,8 +24,11 @@ async def main():
     await kv.set_async("previous_name", name)
 
     # Async HTTP call with httpx (PEP 723 inline dep above).
+    # httpbin_url defaults to https://httpbin.org/get but can be overridden
+    # via params (e.g. by tests using a local httptest server).
+    httpbin_url = await params.get_async("httpbin_url", "https://httpbin.org/get")
     async with httpx.AsyncClient(timeout=5) as client:
-        resp = await client.get("https://httpbin.org/get", params={"name": name})
+        resp = await client.get(httpbin_url, params={"name": name})
         resp.raise_for_status()
         origin = resp.json().get("origin", "unknown")
     await log.info_async(f"Request origin: {origin}")

--- a/tasks/examples/hello-python/task.yaml
+++ b/tasks/examples/hello-python/task.yaml
@@ -17,10 +17,9 @@ params:
   count:
     description: Run count (shown in greeting)
     default: "1"
-  httpbin_url:
-    description: HTTP endpoint to GET (httpbin-compatible JSON response). Override in tests to use a local server.
-    default: "https://httpbin.org/get"
 
 permissions:
   net:
     - httpbin.org
+  env:
+    - HTTPBIN_URL

--- a/tasks/examples/hello-python/task.yaml
+++ b/tasks/examples/hello-python/task.yaml
@@ -17,6 +17,9 @@ params:
   count:
     description: Run count (shown in greeting)
     default: "1"
+  httpbin_url:
+    description: HTTP endpoint to GET (httpbin-compatible JSON response). Override in tests to use a local server.
+    default: "https://httpbin.org/get"
 
 permissions:
   net:

